updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use log::warn;
+use std::{collections::HashMap, fs, path::Path};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "updatehub-statechange.sock";
+
+type StateChangeCallback = Box<dyn Fn(State)>;
+type ErrorCallback = Box<dyn Fn(String)>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    listeners: HashMap<StateID, Vec<StateChangeCallback>>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    pub id: StateID,
+    pub conn: UnixStream,
+}
+
+impl State {
+    // Cancels the current action on the agent.
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    // Tell the agent to proceed with the transition.
+    // No message need to be sent to the connection in order to the
+    // agent to proceed handling the current state.
+    pub fn proceed(&self) {}
+}
+
+impl StateChangeListener {
+    pub fn new() -> Self {
+        StateChangeListener { listeners: HashMap::new(), error_handlers: Vec::new() }
+    }
+
+    pub fn on_state_change(&mut self, state: StateID, f: StateChangeCallback) {
+        self.listeners.entry(state).or_insert_with(|| vec![Box::new(f)]);
+    }
+
+    pub fn on_error(&mut self, f: ErrorCallback) {
+        self.error_handlers.push(f);
+    }
+
+    pub async fn listen(&self) -> Result<()> {
+        let file = Path::new(&SDK_TRIGGER_FILENAME);
+        if !file.exists() {
+            warn!(
+                "WARNING: updatehub-sdk-statechange-trigger not found on {}",
+                SDK_TRIGGER_FILENAME
+            );
+        }
+        fs::remove_file(SOCKET_PATH)?;
+        let listener = UnixListener::bind(SOCKET_PATH).await?;
+
+        loop {
+            match listener.accept().await {
+                Ok((socket, ..)) => self.handle_connection(socket).await,
+                Err(e) => Err(Error::from(e)),
+            }?;
```

Comment:
The Error propagation operator (`?`) can be used here now that we don't need any explicitly typing